src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,549 @@
+		comp_info(dev, "rtnr_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+		ret = comp_data_blob_set_cmd(cd->model_handler, cdata);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_set_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t rtnr_cmd_get_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		comp_dbg(dev, "rtnr_cmd_get_value(), SOF_CTRL_CMD_SWITCH, cdata->comp_id = %u",
+			 cdata->comp_id);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_get_value() error: invalid cdata->cmd %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t rtnr_cmd_set_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		comp_dbg(dev, "rtnr_cmd_set_value(), SOF_CTRL_CMD_SWITCH, cdata->comp_id = %u",
+			 cdata->comp_id);
+		break;
+
+	default:
+		comp_err(dev, "rtnr_cmd_set_value() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int rtnr_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "rtnr_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = rtnr_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = rtnr_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	case COMP_CMD_SET_VALUE:
+		ret = rtnr_cmd_set_value(dev, cdata);
+		break;
+	case COMP_CMD_GET_VALUE:
+		ret = rtnr_cmd_get_value(dev, cdata);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd() error: invalid command");
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int rtnr_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "rtnr_trigger() cmd: %d", cmd);
+
+	return comp_set_state(dev, cmd);
+}
+
+/* copy and process stream data from source to sink buffers */
+static int rtnr_copy(struct comp_dev *dev)
+{
+	struct comp_buffer *sink;
+	struct comp_buffer *source;
+	int frames;
+	int num_blocks;
+	int sink_bytes;
+	int source_bytes;
+	const struct audio_stream *sources_stream[RTNR_MAX_SOURCES] = { NULL };
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "rtnr_copy()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+
+	/* put empty data into output queue*/
+	int nch = source->stream.channels;
+	/* RTKMA uses 32 bits internal storage */
+	int copy_size = BLOCKLENGTH * nch * sizeof(int32_t);
+
+	RTKMA_API_First_Copy((void *)&cd->rtk_agl, copy_size);
+
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	frames = audio_stream_avail_frames(&source->stream, &sink->stream);
+
+	/* Process integer multiple of RTNR internal block length */
+	num_blocks = frames / RTNR_BLK_LENGTH;
+	frames = num_blocks * RTNR_BLK_LENGTH;
```

Comment:
Could you give an introduction of BLOCKLENGTH and RTNR_BLK_LENGTH?
Would RTNR_BLK_LENGTH be a number which is not 2^N? Could we simply this to:
define RTNR_BLK_LENGTH_MASK (RTNR_BLK_LENGTH - 1)
frames = frames & ~RTNR_BLK_LENGTH_MASK

A division and a multiplication operation for each copy looks costy.

Reply:
Yes RTNR_BLK_LENGTH must be power of 2. These lines are simplified. Thanks.